File: structures/Array.cpp
#include "Array.h"
#include <iostream>

MyArray::MyArray() : capacity(2), size(0) {
    data = new int[capacity];
}

MyArray::~MyArray() { delete[] data; }

void MyArray::resize(size_t new_capacity) {
    int* new_data = new int[new_capacity];
    for (size_t i = 0; i < size; i++) { new_data[i] = data[i]; }
    delete[] data;
    data = new_data;
    capacity = new_capacity;
}

void MyArray::add_to_end(int value) {
    if (size == capacity) { resize(capacity * 2); }
    data[size++] = value;
}

void MyArray::add_at_index(size_t index, int value) {
    if (index > size) { throw std::out_of_range("Index out of bounds"); }
    if (size == capacity) { resize(capacity * 2); }
    for (size_t i = size; i > index; i--) { data[i] = data[i - 1]; }
    data[index] = value;
    size++;
}

int MyArray::get_at_index(size_t index) const {
    if (index >= size) { throw std::out_of_range("Index out of bounds"); }
    return data[index];
}

void MyArray::remove_at_index(size_t index) {
    if (index >= size) { throw std::out_of_range("Index out of bounds"); }
    for (size_t i = index; i < size - 1; i++) { data[i] = data[i + 1]; }
    size--;
}

void MyArray::replace_at_index(size_t index, int value) {
    if (index >= size) { throw std::out_of_range("Index out of bounds"); }
    data[index] = value;
}

size_t MyArray::get_length() const { return size; }

void MyArray::print() const {
    std::cout << "Array [";
    for (size_t i = 0; i < size; i++) {
        std::cout << data[i];
        if (i < size - 1) std::cout << ", ";
    }
    std::cout << "]\n";
}

void MyArray::serialize(const std::string& filename) const {
    std::ofstream file(filename, std::ios::binary);
    if (!file) throw std::runtime_error("Cannot open file for writing");
    
    file.write(reinterpret_cast<const char*>(&size), sizeof(size));
    file.write(reinterpret_cast<const char*>(data), size * sizeof(int));
    file.close();
}

void MyArray::deserialize(const std::string& filename) {
    std::ifstream file(filename, std::ios::binary);
    if (!file) throw std::runtime_error("Cannot open file for reading");
    
    size_t new_size;
    file.read(reinterpret_cast<char*>(&new_size), sizeof(new_size));
    
    if (new_size > capacity) { resize(new_size); }
    size = new_size;
    
    file.read(reinterpret_cast<char*>(data), size * sizeof(int));
    file.close();
}

File: Go/coverage/Array.go
package datastructures

import (
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"os"
)

type MyArray struct {
	data     []int
	capacity int
	size     int
}

func NewMyArray() *MyArray {
	return &MyArray{
		data:     make([]int, 2),
		capacity: 2,
		size:     0,
	}
}

func (a *MyArray) resize(newCapacity int) {
	newData := make([]int, newCapacity)
	copy(newData, a.data[:a.size])
	a.data = newData
	a.capacity = newCapacity
}

func (a *MyArray) AddToEnd(value int) {
	if a.size == a.capacity {
		a.resize(a.capacity * 2)
	}
	a.data[a.size] = value
	a.size++
}

func (a *MyArray) AddAtIndex(index int, value int) error {
	if index < 0 || index > a.size  {
		return errors.New("index out of bounds")
	}
	if a.size == a.capacity {
		a.resize(a.capacity * 2)
	}
	for i := a.size; i > index; i-- {
		a.data[i] = a.data[i-1]
	}
	a.data[index] = value
	a.size++
	return nil
}

func (a *MyArray) GetAtIndex(index int) (int, error) {
	if index < 0 || index >= a.size {
		return 0, errors.New("index out of bounds")
	}
	return a.data[index], nil
}

func (a *MyArray) RemoveAtIndex(index int) error {
	if  index < 0 || index >= a.size{
		return errors.New("index out of bounds")
	}
	for i := index; i < a.size-1; i++ {
		a.data[i] = a.data[i+1]
	}
	a.size--
	return nil
}

func (a *MyArray) ReplaceAtIndex(index int, value int) error {
	if index < 0 || index >= a.size  {
		return errors.New("index out of bounds")
	}
	a.data[index] = value
	return nil
}

func (a *MyArray) GetLength() int {
	return a.size
}

func (a *MyArray) Print() {
	fmt.Print("Array [")
	for i := 0; i < a.size; i++ {
		fmt.Print(a.data[i])
		if i < a.size-1 {
			fmt.Print(", ")
		}
	}
	fmt.Println("]")
}

// Binary Serialization
func (a *MyArray) Serialize(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("cannot open file for writing: %w", err)
	}
	defer file.Close()

	if err := binary.Write(file, binary.LittleEndian, uint64(a.size)); err != nil {
		return err
	}
	for i := 0; i < a.size; i++ {
		if err := binary.Write(file, binary.LittleEndian, int32(a.data[i])); err != nil {
			return err
		}
	}
	return nil
}

func (a *MyArray) Deserialize(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return fmt.Errorf("cannot open file for reading: %w", err)
	}
	defer file.Close()

	var newSize uint64
	if err := binary.Read(file, binary.LittleEndian, &newSize); err != nil {
		return err
	}

	if int(newSize) > a.capacity {
		a.resize(int(newSize))
	}
	a.size = int(newSize)

	for i := 0; i < a.size; i++ {
		var value int32
		if err := binary.Read(file, binary.LittleEndian, &value); err != nil {
			return err
		}
		a.data[i] = int(value)
	}
	return nil
}

// JSON Serialization
type arrayJSON struct {
	Data []int `json:"data"`
}

func (a *MyArray) SerializeJSON(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("cannot open file for writing: %w", err)
	}
	defer file.Close()

	data := arrayJSON{Data: a.data[:a.size]}
	encoder := json.NewEncoder(file)
	encoder.SetIndent("", "  ")
	return encoder.Encode(data)
}

func (a *MyArray) DeserializeJSON(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return fmt.Errorf("cannot open file for reading: %w", err)
	}
	defer file.Close()

	var data arrayJSON
	decoder := json.NewDecoder(file)
	if err := decoder.Decode(&data); err != nil {
		return err
	}

	newSize := len(data.Data)     
	if newSize > a.capacity {      
    	a.resize(newSize)         
	}
	a.size = newSize              
	copy(a.data, data.Data)
	return nil
}
